### DISCUSSION TITLE
Java client 5.4.3 frame_max default value on server side

### DISCUSSION THREAD

**Discussion by OlehD** [rabbitmq/rabbitmq-java-client#729] [Q&A] [answered]
After upgrade we started having such issue `Content headers exceeded max frame size: 262560 > 131072`. 

Seems like in the new code we started negotiating for the value to be used basing on the frameMax values from client and server sides. On client side in ConnectionFactory, I found that we have a default value set to zero which means unlimited. However, on server we never set a value, and I couldn't find what is default value there.

Does anyone know if there is default value on the server side for this property?

**Comment by michaelklishin** [comment]
You are leaving out critically important information: what versions have you upgraded *from*.

The default value is right in the error message: `131072`.  It does not mean that you cannot send messages of a greater size,
as message payload is framed (split into multiple payload frames) by both sides.

We recommend not touching that value on any side or at least using `131072`.

**Reply by OlehD** [reply]
it was upgraded from 3.5.4 to 5.4.3.

actually we can't send a message because it throws exception. here is the code where it is thrown:
```
                if (cappedFrameMax && headerFrame.size() > frameMax) {
                    String msg = String.format("Content headers exceeded max frame size: %d > %d", headerFrame.size(), frameMax);
                    throw new IllegalArgumentException(msg);
                }
```

and this is how we calculate it in 5.4.3:
```
            int frameMax =
                negotiatedMaxValue(this.requestedFrameMax,
                                   connTune.getFrameMax());

...............................

    private static int negotiatedMaxValue(int clientValue, int serverValue) {
        return (clientValue == 0 || serverValue == 0) ?
            Math.max(clientValue, serverValue) :
            Math.min(clientValue, serverValue);
    }

```

We always counted for unlimited value, and with upgrade it seems like client started using default value form server side?...

**Reply by michaelklishin** [reply]
@OlehD just take out the setting on both ends. Why do you have to override it?

**Reply by michaelklishin** [reply]
The logic that picks a number based on two different settings has changed once or twice in RabbitMQ's history. Unless a specific client does it differently from all other clients, we won't change it again. But it's a setting no one needs to set explicitly. I would
take it out of the protocol if I could at this point (but it's obviously too late 15 years in).

**Reply by OlehD** [reply]
because in the end I would like to have frameMax set to 0. So, if I don't miss anything I need negotiatedMaxValue(...) to return me zero, and it is possible only when client and server values are both 0.

I understand that it is not preferable to have unlimited size, but application is working in this way, and refactoring is not quick thing to do...

**Reply by OlehD** [reply]
I didn't check yet how it was working in version 3.5.4 Once I check then I will update here. Might be the case when it was changed in between 3.5.4 and 5.4.3

**Reply by michaelklishin** [reply]
I still think that it doesn't matter because you should simply take that option out on both ends. That's how nearly all Java client test suites run, and most RabbitMQ test suites as well. We don't override max frame size, there is no reason to (other than to test protocol framing).

**Reply by OlehD** [reply]
we are doing some testing right now. I will update here with results once I will have something.

**Reply by OlehD** [reply]
if i understand the logic correctly then it will be impossible to set unlimited size unless it is set explicitly on both sides client and server.

**Reply by michaelklishin** [reply]
Why do you need it to be unlimited in the first place? You can publish messages up to 128 MB in size in modern RabbitMQ versions,
and even larger (up to 512 MiB?) with some configuration. No one should do that, however, put those large blobs into a blob store and pass their IDs around.

Every message larger than 128 kiB will be framed by the client. So, a `basic.publish` frame will be followed by a content (metadata) frame and one or more body frames.

**Reply by OlehD** [reply]
even if we set client with 128MB the still we need to set server to 0, otherwise with current logic it will pick up min value. 

```
return (clientValue == 0 || serverValue == 0) ?
            Math.max(clientValue, serverValue) :
            Math.min(clientValue, serverValue);
```

btw, when I set frame max to zero on server side then all worked. The current logic basically limits the client to the value on the server. So, in any case we need to set it on server side.

Yes, agree that this question is more for java client side. Thanks for moving it!

**Reply by michaelklishin** [reply]
@OlehD we are running in circles. You don't need to override this value. *At all*. Even if your messages are larger than 128 kiB in size. It's an optional setting, just un set it everywhere and let RabbitMQ and client use the default.

**Reply by michaelklishin** [reply]
A client cannot be allowed to tell the server what value to use. Our team has learnt this the hard way. Issues that come down to client abuse (deliberate or just sloppy coding or unfortunate configuration values) are not uncommon at all.

**Comment by michaelklishin** [comment]
Transferred this discussion to the Java client repo as it seems more appropriate here at this point.

**Comment by michaelklishin** [comment, marked as answer]
@OlehD we are running in circles. You don't need to override this value. *At all*. Even if your messages are larger than 128 kiB in size. It's an optional setting, just unset it everywhere and let RabbitMQ and client use the default. 

If your messages are into 10s of megabytes, and you transfer them using RabbitMQ, then it will technically work but you are doing it wrong. Every part of storage was not designed with large messages in mind, although streams likely will handle them better than any of the queue types. Put them in a blob store and pass their IDs or URIs around.

We won't change the logic to allow the client to override server value. That's a terrible idea and I'm sure you can guess why.

**Reply by OlehD** [reply]
Do I understand you correctly with this statement - "unset it everywhere and let RabbitMQ and client use the default" - you propose to remove the property frame_max from rabbitmq.conf, and don't set it in ConnectionFactory? If so it was not set in those two places before upgrade. 

I believe what happened is that in 3.5.4 it was not working as expected on client side, and in higher versions something was fixed on client side, and it started working as expected since then. However, it requires adjustments in settings when someone does upgrade, and uses big messages in size.... I still try to figure out why upgrade caused it, and if I find something then I will post it here. Also, I agree that it is not the best practice to send big messages, but it is not an easy/quick thing to refactor old legacy code in a big service...

I'm not asking to make changes on client side, but I just wanted to understand why it needed now, so I could make a proper fix in configs for this. Also, maybe it would be good to add something into the specification about how exactly to set up frame max property?

Just to mention that the same issue happened with channel max after upgrade.

**Reply by michaelklishin** [reply]
Yes, I mean that you don't override the value on either end.

You are the first one to report this. [5.4.3](https://github.com/rabbitmq/rabbitmq-java-client/releases/tag/v5.4.3) shipped in
October 2018, more than three years ago. The only relevant change in the last three years that I recall was
the introduction of a hard limit on message size and a lower default (128 MiB) for it.

https://github.com/rabbitmq/rabbitmq-java-client/blame/27203011bd60611af870cc5e092d838dbfe0f3cb/src/main/java/com/rabbitmq/client/impl/AMQConnection.java#L409-L414 demonstrates that the key method in question was last updated 11 years ago.

You can begin using any later 5.x version of the client today, with whatever RabbitMQ version you run. This is not worth
any time spent on code archeology if you ask me.

**Reply by michaelklishin** [reply]
The only relevant change in RabbitMQ itself that I could find (ignoring `frame_max` uses in the stream protocol, which this client does not use) is this from 2020: https://github.com/rabbitmq/rabbitmq-server/commit/157808ca8fa08346b02c5ce9747880009ae5442c.

I don't see how it can be relevant since neither the original nor the new value are `0`.

**Reply by michaelklishin** [reply]
Err, I wanted to link to https://github.com/rabbitmq/rabbitmq-java-client/blame/27203011bd60611af870cc5e092d838dbfe0f3cb/src/main/java/com/rabbitmq/client/impl/AMQConnection.java#L633-L636 but the way frame max negotiation method is called is also relevant. Anyhow, this time the changes were 7 and 14 years ago.

So I don't see any evidence of meaningful changes since October 2018. You may be misinterpreting the behavior, what settings
were used at what point in your project's history, and so on.

Just upgrade the client, drop the overrides and move on.

**Reply by michaelklishin** [reply]
https://github.com/rabbitmq/rabbitmq-java-client/commit/733788e26ba73ea90a1be72b99227c6a10003054 and https://github.com/rabbitmq/rabbitmq-java-client/issues/640 can also be relevant, they happened after `5.4.3` (in early 2020).

That shipped in `5.8.1`.
